cl/beacon/validatorapi/post.go

```diff
@@ -46,14 +52,101 @@ func (v *ValidatorApiHandler) PostEthV1ValidatorSyncCommitteeSubscriptions(w htt
 	return nil, beaconhttp.NewEndpointError(404, errNotImplemented)
 }
 
-func (v *ValidatorApiHandler) PostEthV1ValidatorBeaconCommitteeSubscriptions(w http.ResponseWriter, r *http.Request) (*int, error) {
+func (v *ValidatorApiHandler) PostEthV1ValidatorBeaconCommitteeSubscriptions(w http.ResponseWriter, r *http.Request) (*string, error) {
 	var req []building.BeaconCommitteeSubscription
 	err := json.NewDecoder(r.Body).Decode(&req)
 	if err != nil {
 		return nil, beaconhttp.NewEndpointError(400, fmt.Errorf("invalid request: %w", err))
 	}
-	// TODO: this endpoint
-	return nil, beaconhttp.NewEndpointError(404, errNotImplemented)
+
+	// the empty list i believe is a valid request, and it will just do nothing
+	if len(req) == 0 {
+		return nil, nil
+	}
+
+	l := log.New()
+
+	/*
+		After beacon node receives this request,
+		search using discv5 for peers related to this subnet and replace current peers with those ones if necessary.
+		If validator is_aggregator, beacon node must:
+
+		 1. announce subnet topic subscription on gossipsub
+		 2. aggregate attestations received on that subnet
+	*/
+
+	// grab the head state to calculate committee counts with
+	s, err := v.privateGetStateFromStateId("head")
+	if err != nil {
+		return nil, err
+	}
+
+	digest, err := fork.ComputeForkDigest(v.BeaconChainCfg, v.GenesisCfg)
+	if err != nil {
+		l.Error("[Gossip] Failed to calculate fork choice", "err", err)
+	}
+
+	//calculate gossip_subnets
+	topics := make([]string, 0, len(req))
+	for _, x := range req {
+		// skip ones we are aggregator for
+		if !x.IsAggregator {
+			continue
+		}
+		committees_per_slot := s.CommitteeCount(uint64(x.Slot) / v.BeaconChainCfg.SlotsPerEpoch)
+		subnet := v.BeaconChainCfg.ComputeSubnetForAttestation(committees_per_slot, uint64(x.Slot), uint64(x.CommitteeIndex))
+		topicString := fmt.Sprintf("/eth2/%x/beacon_attestation_%d/%s", digest, subnet, "ssz_snappy")
+		topics = append(topics, topicString)
+	}
+
+	filterString := "/eth2/*/beacon_attestation_*/ssz_snappy"
+	subscriptionData := &sentinel.SubscriptionData{
+		Filter: &filterString,
+		Topics: topics, // TODO: this should be done in advance in the sentinel and recalculated every once in a while
+	}
+	sub, err := v.Sentinel.SubscribeGossip(r.Context(), subscriptionData)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		l.Info("[beacon api] started aggregating gossip", "subscription", subscriptionData)
+		defer func() {
+			l.Info("[beacon api] aggregation gossip subscriber closed")
+		}()
+		for {
```

Comment:
i think that the subnet listening is supposed to expire eventually. also make sure it does not subscribe twice